src/codegen/spirv.zig

```diff
@@ -1774,29 +1817,28 @@ const DeclGen = struct {
         }
 
         /// Utility function to extract the element at a particular index in an
-        /// input vector. This type is expected to be a vector if `wip.is_vector`, and
-        /// a scalar otherwise.
+        /// input array. This type is expected to be a fake vector (array) if `wip.is_array`, and
+        /// a vector or scalar otherwise.
         fn elementAt(wip: WipElementWise, ty: Type, value: IdRef, index: usize) !IdRef {
             const mod = wip.dg.module;
-            if (wip.is_vector) {
+            if (wip.is_array) {
                 assert(ty.isVector(mod));
                 return try wip.dg.extractField(ty.childType(mod), value, @intCast(index));
             } else {
-                assert(!ty.isVector(mod));
                 assert(index == 0);
                 return value;
             }
         }
 
-        /// Turns the results of this WipElementWise into a result. This can either
-        /// be a vector or single element, depending on `result_ty`.
+        /// Turns the results of this WipElementWise into a result. This can be
+        /// vectors, fake vectors (arrays) and single elements, depending on `result_ty`.
         /// After calling this function, this WIP is no longer usable.
         /// Results is in `direct` representation.
         fn finalize(wip: *WipElementWise) !IdRef {
-            if (wip.is_vector) {
+            if (wip.is_array) {
                 // Convert all the constituents to indirect, as required for the array.
                 for (wip.results) |*result| {
-                    result.* = try wip.dg.convertToIndirect(wip.scalar_ty, result.*);
+                    result.* = try wip.dg.convertToIndirect(wip.ty, result.*);
```

Comment:
The results of finalize() should be direct representation, so I think this should just call constructComposite directly.

Reply:
Or did you not yet implement that part?

Reply:
I didn't, I was afraid that it will get big and complicated like the last two attempts

Reply:
ok, we'll do it in a different pr